bicytok/distanceMetricFuncs.py

```diff
@@ -1,641 +1,221 @@
+
+    return targAbun, offTargAbun
+
+
+def calculateKLDiv(targAbun, offTargAbun, dims):
     """
-    conversion factors used for citeseq dataset
-    :param weightDF: DataFrame containing weight information for receptors
-    :param receptor_name: Name of the receptor for which the conversion factor is needed
+    Calculates KL Divergence between target and off-target abundance distributions
+    :param targAbun: in 1D - a vector of target abundance values for one receptor
+        in 2D - an nx2 array where each column are the target abundance values for one receptor
+    :param offTargAbun: in 1D - a vector of off-target abundance values for one receptor
+        in 2D - an nx2 array where each column are the off-target abundance values for one receptor
+    :param dims: the number of receptors being compared across
     :return:
-        conversion_factor: Conversion factor for the specified receptor
+        KL_div_val: the KL Divergence between target and off-target abundances
     """
-    IL2Rb_factor = weightDF.loc[weightDF["Receptor"] == "IL2Rb", "Weight"].values[0]
-    IL7Ra_factor = weightDF.loc[weightDF["Receptor"] == "IL7Ra", "Weight"].values[0]
-    IL2Ra_factor = weightDF.loc[weightDF["Receptor"] == "IL2Ra", "Weight"].values[0]
-    if receptor_name == "CD122":
-        return IL2Rb_factor
-    elif receptor_name == "CD25":
-        return IL2Ra_factor
-    elif receptor_name == "CD127":
-        return IL7Ra_factor
-    else:
-        return (IL7Ra_factor + IL2Ra_factor + IL2Rb_factor) / 3
-
-
-def EMD_2D(dataset, signal_receptor, target_cells, special_receptor, ax):
+
+    if dims == 1:
+        targKDE = KernelDensity(kernel="gaussian").fit(targAbun.reshape(-1, 1))
+        offTargKDE = KernelDensity(kernel="gaussian").fit(offTargAbun.reshape(-1, 1))
+        minAbun = np.minimum(targAbun.min(), offTargAbun.min()) - 10
+        maxAbun = np.maximum(targAbun.max(), offTargAbun.max()) + 10
+        outcomes = np.arange(minAbun, maxAbun + 1).reshape(-1, 1)
+
+    elif dims == 2:  # LOOK FOR WAY TO GENERALIZE FOR ALL DIMS
+        targKDE = KernelDensity(kernel="gaussian").fit(targAbun.reshape(-1, 2))
+        offTargKDE = KernelDensity(kernel="gaussian").fit(offTargAbun.reshape(-1, 2))
+        minAbun = np.minimum(targAbun.min(), offTargAbun.min()) - 10
+        maxAbun = np.maximum(targAbun.max(), offTargAbun.max()) + 10
+        X, Y = np.mgrid[
+            minAbun : maxAbun : ((maxAbun - minAbun) / 100),
+            minAbun : maxAbun : ((maxAbun - minAbun) / 100),
+        ]
+        outcomes = np.concatenate((X.reshape(-1, 1), Y.reshape(-1, 1)), axis=1)
+
+    targDist = np.exp(targKDE.score_samples(outcomes))
+    offTargDist = np.exp(offTargKDE.score_samples(outcomes))
+    KL_div_val = stats.entropy(
+        offTargDist.flatten() + 1e-200, targDist.flatten() + 1e-200, base=2
+    )
+
+    return KL_div_val
+
+
+def KL_EMD_1D(
+    recAbundances: np.ndarray, targ: np.ndarray, offTarg: np.ndarray
```

Comment:
Could you add typing like this to the other functions? https://numpy.org/doc/stable/reference/typing.html

Reply:
Done